src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,132 @@
+    private int size = 0;
+
+    public ArrayList() {
+        elementData = DEFAULTCAPACITY_EMPTY_ELEMENTDATA;
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == elementData.length) {
+            elementData = grow();
+        }
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index != size) {
+            rangeCheck(index);
+        }
+        if (size >= elementData.length) {
+            elementData = grow();
+        }
+        System.arraycopy(this.elementData, index, this.elementData, index + 1, size - index);
+        this.elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        rangeCheck(index);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        rangeCheck(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        rangeCheck(index);
+        final Object[] es = elementData;
+        T oldValue = (T) es[index];
+        fastRemove(es, index);
+        return oldValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        final Object[] es = elementData;
+        final int size = this.size;
+        int i = 0;
+        found: {
+            if (element == null) {
+                for (; i < size; i++) {
+                    if (es[i] == null) {
+                        break found;
+                    }
+                }
+            } else {
+                for (; i < size; i++) {
+                    if (element.equals(es[i])) {
+                        break found;
+                    }
+                }
+            }
+            throw new NoSuchElementException();
+        }
+        fastRemove(es, i);
+        return element;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void rangeCheck(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(outOfBoundsMsg(index));
+        }
+    }
+
+    private String outOfBoundsMsg(int index) {
+        return "Index: " + index + "out of bounds, cause size is: " + size;
+    }
+
+    private void fastRemove(Object[] es, int i) {
+        final int newSize;
+        if ((newSize = size - 1) > i) {
+            System.arraycopy(es, i + 1, es, i, newSize - i);
+        }
+        es[size = newSize] = null;
+    }
+
+    private Object[] grow() {
+        return grow(size + 1);
+    }
+
+    private Object[] grow(int minCapacity) {
```

Comment:
why do you need 2 methods? 
```suggestion
    private void grow() {
```